Add carbon emissions reporting to job information page

Renni771 opened the pull request:
Addresses #15046. This PR adds an initial implementation of carbon emissions reporting to the job information page. It currently assumes global average values for carbon intensity and power usage effectiveness and does not take the user's or server's location into account. The next step would be to dynamically determine the CPU configuration and location of the server that ran a specific job to allow for more accurate estimates.

Screenshots of the new UI:

The new UI using dummy data:
![Screenshot from 2023-05-31 11-17-02](https://github.com/galaxyproject/galaxy/assets/57318507/68491269-8de6-4a59-81c4-2db4b6feb222)

Info can be accessed on hover and click:
[Screencast from 2023-05-31 12-05-02.webm](https://github.com/galaxyproject/galaxy/assets/57318507/abedf9b7-36af-47f1-a63b-17ec83699449)

A new route explaining how we make the calculations was added:
[Screencast from 2023-05-31 11-17-57.webm](https://github.com/galaxyproject/galaxy/assets/57318507/adc9a76d-9ddd-4f1c-9934-cca0bdc1ef08)

Notes:
- The graph only shows when there is more than one metric to show. i.e, if we only get cpu info from the job metrics endpoint, only the table shows. The table centers itself in this case to better fill the white space.
- A threshold can be set for each `CarbonEmissionsCard` component - these are the tiles on the left of the table and graph. When the value is below the threshold, we say the value is less than or equal to the threshold so as to avoid tiny decimal numbers or values that don't really make sense for the comparison.

Other changes:
- The AWS estimates is now its own `AwsEstimates` component. The component itself was refactored 
- The AWS estimates and `JobMetrics`  tests were refactored to support the new component splitting.
- The `ec2.json` contained reference data explaining how various AWS EC2 machines were configured. This was used to determine the AWS estimates. This file was updated and converted to a `.js` file with more up-to-date values listed in the AWS EC2 docs. Since the file is large, we dynamically import it on demand and pass its data as props to the `AwsEstimates` component which needs the same data. For now this same EC2 data is used to estimate the configuration of a server instance that could have run the current job. This is the final fallback when we have little or no information from the server about the server the ran the job. 
- An `Abbreviation` component was added.
- The `Heading` component was converted to typescript.
- A `GetComponentPropTypes` utility type was added.


## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

Renni771 committed:
Add carbon emssions reporting to job information page

Renni771 committed:
Make memory allocated nullable and correctly calculated normalized TDP
per core

Renni771 committed:
Adjust JobMetrics component tests and add AwsEstimate component tests

Renni771 committed:
Determine server CPU info in JobMetrics component and add units to props

- Users of the CarbonEmissions component should pass all necessary
  props
- Add explicit units to prop names

Renni771 committed:
Add unit tests for CarbonEmissions component

Renni771 committed:
Add units to AwsEstimate component props and format code

Renni771 committed:
Remove unused carbon emissions constants file

Renni771 committed:
Remove unused import

Renni771 committed:
Make equation blocks stand out more

Renni771 committed:
Improve grammar, fix typos and clean up code

bgruening commented on client/package.json:
```diff
@@ -43,6 +43,7 @@
     "bootstrap": "4.6",
     "bootstrap-vue": "^2.23.0",
+    "chart.js": "4.3.0",
     "citation-js": "^0.6.4",
```
Can we use d3.js here and avoid this extra dependencie?

Renni771 replied:
Sure, I've swapped it out for D3.

Renni771 committed:
Use D3 for carbon emission usage graph

Renni771 committed:
Colocate carbon emissions and Abbreviation components

* Move carbon emissions UI to job metrics directory
* Move Abbreviation component to Common

Renni771 committed:
Merge branch 'galaxyproject:dev' into carbon-emissions-reporting

Renni771 committed:
Carbon emissions code clean up and minor bug fixes

Renni771 committed:
Convert Heading component to use typescript

Renni771 committed:
Use font awesome icons in carbon comparisons and add icon type safety

Renni771 committed:
Remove hover effects and adjust style to fit branding

Renni771 committed:
Only show bar graph when more than one metric can be shown

Renni771 committed:
Lazy load AWS EC2 data

Renni771 committed:
Add sources to AWS reference data

Renni771 committed:
Fix CPU power consumption calculation by using cores allocated to job

Renni771 committed:
Show `CarbonEmissionsCard` explanations on click

Renni771 committed:
Format code

Renni771 committed:
Compare energy usage to number of light bulbs running

Renni771 committed:
Reword explanations

Renni771 committed:
Remove exported prop type interfaces in JobMetrics and CarbonEmissions components

Renni771 committed:
Add a utility type to extract the prop types of a component

ElectronicBlueberry committed:
several tweaks and fixes
make bar chart reactive and use ref for multiple instances of component
typo fixes
remove unused nesting from CarbonEmissionsIcon
remove required from JobId

ElectronicBlueberry committed:
fix gasoline consumption calculation
tweak lightbulb energy description

ElectronicBlueberry reviewed (approved):
Thanks for all the work on this, looks and works great!

Renni771 committed:
Always show "distance driven" metrics next to each other
